// Copyright 2023 Intrinsic Innovation LLC

syntax = "proto3";

package intrinsic_proto.skills;

import "google/protobuf/empty.proto";
import "intrinsic/resources/proto/resource_handle.proto";
import "intrinsic/skills/proto/skill_registry_config.proto";
import "intrinsic/skills/proto/skills.proto";

message GetInstanceRequest {
  // The skill's id. This can be either the id or the id_version.
  string id = 4;

  // An optional instance id for the skill. If empty, a unique id will be
  // generated by the registry.
  string instance_id = 3;

  // The resource the instance will use.
  map<string, intrinsic_proto.resources.ResourceHandle> handles = 2;

  reserved 1;
}

message GetInstanceResponse {
  // The returned instance. The contents of this field will only be valid when
  // `status` is OK.
  intrinsic_proto.skills.SkillInstance instance = 2;

  reserved 1;
  reserved "status";
}

message RemoveSkillRequest {
  // The id of the skill to remove.
  string id = 1;
}

message RemoveSkillResponse {}

message RegisterOrUpdateSkillRequest {
  // The registration information for the skill.
  intrinsic_proto.skills.SkillRegistration skill_registration = 1;
}

service SkillRegistryInternal {
  // Returns a skill instance matching the given request.
  rpc GetInstance(GetInstanceRequest) returns (GetInstanceResponse) {}

  // Removes a skill from the registry. This method is not required to be
  // implemented by all implementations of the skill registry.
  rpc RemoveSkill(RemoveSkillRequest) returns (RemoveSkillResponse) {}

  // This is deprecated and currently only works for in-memory implementations
  // (typically used for unit testing). Registers or updates a skill.
  rpc RegisterOrUpdateSkill(RegisterOrUpdateSkillRequest)
      returns (google.protobuf.Empty) {}

  // Resets the set of instance names.
  //
  // All skill execution operations must have finished before calling this RPC,
  // since it also calls each unique skill executor service's ClearOperations().
  rpc ResetInstanceIDs(google.protobuf.Empty) returns (google.protobuf.Empty) {}
}
